flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -600,8 +600,14 @@ private ChangelogSnapshotState completeRestore(
                                     System.currentTimeMillis(),
                                     streamFactory,
                                     CHECKPOINT_OPTIONS),
-                            // TODO: add metadata to log FLINK-23170.
-                            upTo));
+                            upTo);
+
+            // log metadata after materialization is triggered
+            keyValueStatesByName
+                    .values()
+                    .forEach(t -> ((AbstractChangelogState) t).resetMetaDataWritten());
```

Comment:
1. Please use `ChangelogState` instead of `AbstractChangelogState`
2. Please add a call  for `priorityQueueStatesByName` (item above will force to implement it for PQ states)
3. Although this doesn't happen very often, it blocks processing, so I'd rather use for-loop

Reply:
Can you be more explicit what is the difference between `forEach` and `for-loop` here?

Why it is blocking?

Reply:
Sure, sorry, I mean the whole function is executed by the main task thread and this is blocking normal processing.
And `forEach` can be slower than a simple `for` loop, so it makes sense to shorten this time.

Reply:
1. Is it for sure that `forEach` is slower than `for` loop in java? I've searched a bit and found contradict conclusions:
https://answers-short.com/en/Q%26A/page=48105dc27b0dd6653125b9bc3aa0257a

2. Even it is true, given there are very limited number of states, an't the differences really negligible?

Reply:
I think the page you are referring to compares indexed loops with java-1.5 foreach loops, but not with functional `forEach` from  java 8 (and it's also not backed by any benchmarks).
In general, 1.8 loops are slower because of lambdas and function calls.

Yes, it may be neglibible, or may not. I don't think we can assume that it will,;and benchmarking seems non-trivial. So I'd just stay with faster for-loop.